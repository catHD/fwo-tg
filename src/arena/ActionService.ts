import arena from '@/arena';
import type { Action } from '@fwo/schemas';

export type ActionKey = keyof typeof arena.actions;

export class ActionService {
  static isAction(maybeAction: string): maybeAction is ActionKey {
    return maybeAction in arena.actions;
  }

  static isBaseAction(action: string) {
    if (ActionService.isAction(action)) {
      return arena.actions[action].lvl === 0;
    }

    return false;
  }

  static toObject(action: ActionKey): Action {
    const c = arena.actions[action];

    if ('cost' in c && 'costType' in c) {
      return {
        name: c.name,
        displayName: c.displayName,
        costType: c.costType,
        cost: c.cost,
        orderType: c.orderType,
      };
    }

    return {
      name: c.name,
      displayName: c.displayName,
      orderType: c.orderType,
    };
  }
}
