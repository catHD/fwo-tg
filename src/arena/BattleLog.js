const ee = require('events');
const { weaponTypes } = require('./MiscService');

/**
 * msg
 * @todo WIP, —Ñ—É–Ω–∫—Ü–∏—è –¥–æ–ª–∂–Ω–∞ –±—É–¥–µ—Ç –ø—Ä–∏–Ω–∏–º–∞—Ç—å –∫–∞–∫ –∑–Ω–∞—á–µ–Ω–∏—è —É—Ä–æ–Ω–∞ —Ç.–ø —Ç–∞–∫ –∏
 * —É–º–µ—Ç—å —Ä–∞–±–æ—Ç–∞—Ç—å —Å i18n
 * —Å–µ–π—á–∞—Å (–µ) –Ω–µ –æ–±—Ä–∞–±–∞—Ç—ã–≤–∞–µ—Ç—Å—è, –Ω—É–∂–Ω–æ –æ–±—Ä–∞–±–∞—Ç—ã–≤–∞—Ç—å —Ç–æ–ª—å–∫–æ –Ω—É–∂–Ω—ã–π —Ç–∏–ø Error
 * –µ—Å–ª–∏ —ç—Ç–æ –Ω–µ BattleLog –≤—ã–±—Ä–æ—Å, –µ–≥–æ –Ω—É–∂–Ω–æ –ø—Ä–æ–∫–∏–¥—ã–≤–∞—Ç—å –¥–∞–ª—å—à–µ –≤–≤–µ—Ä—Ö
 * @return {String} ({type:String,message:String})
 */
function csl(msgObj) {
  const {
    action, message, target, initiator, expArr, weapon,
  } = msgObj;

  const expString = expArr ? expArr.map(([name, exp]) => `${name}: üìñ${exp}`).join(', ') : '';

  const TEXT = {
    NO_MANA: {
      ru: `–ù–µ —Ö–≤–∞—Ç–∏–ª–æ –º–∞–Ω—ã –¥–ª—è –∑–∞–∫–ª–∏–Ω–∞–Ω–∏—è _${action}_`,
      eng: '',
    },
    NO_ENERGY: {
      ru: `–ù–µ —Ö–≤–∞—Ç–∏–ª–æ —ç–Ω–µ—Ä–≥–∏ –¥–ª—è —É–º–µ–Ω–∏—è _${action}_`,
      eng: '',
    },
    SILENCED: {
      ru: `*${initiator}* –ø—ã—Ç–∞–ª—Å—è —Å–æ—Ç–≤–æ—Ä–∏—Ç—å _${action}_, –Ω–æ –ø–æ–ø—ã—Ç–∫–∞ –ø—Ä–æ–≤–∞–ª–∏–ª–∞—Å—å(–∑–∞—Ç—ã–∫–∞)`,
      eng: '',
    },
    CHANCE_FAIL: {
      ru: `*${initiator}* –ø—ã—Ç–∞–ª—Å—è —Å–æ—Ç–≤–æ—Ä–∏—Ç—å _${action}_, –Ω–æ —É –Ω–µ–≥–æ –Ω–µ –≤—ã—à–ª–æ`,
      eng: '',
    },
    GOD_FAIL: {
      ru: `–ó–∞–∫–ª–∏–Ω–∞–Ω–∏–µ _${action}_ *${initiator}* –ø—Ä–æ–≤–∞–ª–∏–ª–æ—Å—å –ø–æ –≤–æ–ª–µ –±–æ–≥–æ–≤ `,
      eng: '',
    },
    HEAL_FAIL: {
      ru: `*${initiator}* –ø—ã—Ç–∞–ª—Å—è _–≤—ã–ª–µ—á–∏—Ç—å_ *${target}*, –Ω–æ —Ç–æ—Ç –±—ã–ª –∞—Ç–∞–∫–æ–≤–∞–Ω`,
      eng: '',
    },
    SKILL_FAIL: {
      ru: `*${initiator}* –ø—ã—Ç–∞–ª—Å—è –∏—Å–ø–æ–ª—å–∑–æ–≤–∞—Ç—å —É–º–µ–Ω–∏–µ _${action}_, –Ω–æ —É –Ω–µ–≥–æ –Ω–µ –≤—ã—à–ª–æ`,
      eng: '',
    },
    NO_WEAPON: {
      ru: `*${initiator}* –ø—ã—Ç–∞–ª—Å—è –∞—Ç–∞–∫–æ–≤–∞—Ç—å *${target}*, –Ω–æ —É –Ω–µ–≥–æ –Ω–µ –æ–∫–∞–∑–∞–ª–æ—Å—å –æ—Ä—É–∂–∏—è –≤ —Ä—É–∫–∞—Ö`,
      eng: '',
    },
    DEF: {
      ru: `*${initiator}* –∞—Ç–∞–∫–æ–≤–∞–ª *${target}* _${weapon ? weapon.case : ''}_, –Ω–æ —Ç–æ—Ç —Å–º–æ–≥ –∑–∞—â–∏—Ç–∏—Ç—å—Å—è \\[${expString}]`,
      eng: '',
    },
    DODGED: {
      ru: `*${initiator}* –∞—Ç–∞–∫–æ–≤–∞–ª *${target}* _${weapon ? weapon.case : ''}_, –Ω–æ —Ç–æ—Ç —É–∫–ª–æ–Ω–∏–ª—Å—è –æ—Ç –∞—Ç–∞–∫–∏`,
      en: '',
    },
  };
  const text = TEXT[message] || {
    ru: '–û—à–∏–±–∫–∞ –ø–∞—Ä—Å–∏–Ω–≥–∞ —Å—Ç—Ä–æ–∫–∏ –º–∞–≥–∏–∏',
  };
  // @todo —Å–µ–π—á–∞—Å battleLog –Ω–∞ —Å—Ç–æ—Ä–æ–Ω–µ –∫–ª–∏–µ–Ω—Ç–∞ –Ω–µ –ø–æ–Ω–∏–º–∞–µ—Ç —Ç–∏–ø—ã –º–∞–≥–∏–π, –∏
  // –ø—Ä–æ—Å—Ç–æ –æ—Ç–æ–±—Ä–∞–∂–∞–µ—Ç –æ–ø–ª—É—á–µ–Ω–Ω—É—é —Å—Ç—Ä–æ–∫—É
  return text.ru;
}

/**
 * –ö–ª–∞—Å—Å –≤—ã–≤–æ–¥–∞ –¥–∞–Ω–Ω—ã—Ö –≤ battlelog
 * @todo WIP –∫–ª–∞—Å—Å –≤ —Å—Ç–∞–¥–∏–∏ —Ñ–æ—Ä–º–∏—Ä–æ–≤–∞–Ω–∏—è
 * @see https://trello.com/c/qxnIM1Yq/17
 */
class BattleLog extends ee {
  /**
   * –ö–æ–Ω—Å—Ç—Ä—É–∫—Ç–æ—Ä –æ–±—å–µ–∫—Ç–∞ –≤–µ–¥—É—â–µ–≥–æ –ª–æ–≥–∏–∫—É –æ—Ç–¥–∞—á–∏ —Å–æ–æ–±—â–µ–Ω–∏–π –¥–ª—è –ø–æ–ª—å–∑–æ–≤–∞—Ç–µ–ª—å—Å–∫–æ–≥–æ
   * BattleLog
   */
  constructor() {
    super();
    this.msgArray = [];
  }

  /**
   * –§—É–Ω–∫—Ü–∏—è –ª–æ–≥–∏—Ä—É–µ—Ç –¥–µ–π—Å—Ç–≤–∏—è –≤ console log
   * @param {Object.<string, string>} msgObj —Ç–∏–ø —Å–æ–æ–±—â–µ–Ω–∏—è
   */
  log(msgObj) {
    const data = csl(msgObj);
    this.write(data);
  }

  /**
   * –£–¥–∞—á–Ω—ã–π –ø—Ä–æ—Ö–æ–¥ action
   * @param {Object.<string, any>} msgObj —Ç–∏–ø —Å–æ–æ–±—â–µ–Ω–∏—è
   */
  success(msgObj) {
    let data = '';
    const { expArr } = msgObj;
    const expString = expArr ? expArr.map(([name, exp, val]) => `${name}: ‚ù§${val} üìñ${exp}`).join(', ') : '';
    // –ï—Å–ª–∏ –æ–±—å–µ–∫—Ç —Å–æ–¥–µ—Ä–∂–∏—Ç –∫–∞—Å—Ç–æ–º–Ω—É—é —Å—Ç—Ä–æ–∫—É –∏—Å–ø–ª—å–∑—É–µ–º –µ—ë
    if (msgObj.msg) {
      data = msgObj.msg(msgObj.initiator, msgObj.exp);
    } else if (msgObj.dmgType && msgObj.dmgType === 'phys') {
      const { action } = weaponTypes[msgObj.weapon.wtype];
      data = `*${msgObj.initiator}* ${action(msgObj.target, msgObj.weapon)} –∏ –Ω–∞–Ω—ë—Å *${msgObj.dmg}* —É—Ä–æ–Ω–∞ \\[ üíî-${msgObj.dmg}/${msgObj.hp} üìñ${msgObj.exp} ]`;
    } else if (msgObj.dmgType) {
      data = `*${msgObj.initiator}* —Å–æ—Ç–≤–æ—Ä–∏–ª _${msgObj.action}_ (${msgObj.actionType}) –Ω–∞ *${msgObj.target}* –Ω–∞–Ω–µ—Å—è ${msgObj.dmg}  \\[ üíî-${msgObj.dmg}/${msgObj.hp} üìñ${msgObj.exp} ]`;
    } else if (!msgObj.effect) {
      data = `*${msgObj.initiator}* –∏—Å–ø–æ–ª—å–∑–æ–≤–∞–ª _${msgObj.action}_ (${msgObj.actionType}) –Ω–∞ *${msgObj.target}* –∏ –ø–æ–ª—É—á–∏–ª +e:${msgObj.exp}`;
    } else {
      data = `*${msgObj.initiator}* –∏—Å–ø–æ–ª—å–∑–æ–≤–∞–≤ _${msgObj.action}_ –Ω–∞ *${msgObj.target}* —Å —ç—Ñ—Ñ–µ–∫—Ç–æ–º ${msgObj.effect} –ø–æ–ª—É—á–∏–ª +e:${msgObj.exp}`;
    }
    // –í—ã–Ω–æ—à—É –≤–Ω–∏–∑ —Ç.–∫ –ø—Ä–æ–≤–µ—Ä–∫–∞ —Å–≤—è–∑–∞–Ω–∞ —Å action
    if (msgObj.action === 'handsHeal') {
      data = `–ò–≥—Ä–æ–∫ *${msgObj.target}* –±—ã–ª ü§≤ –≤—ã–ª–µ—á–µ–Ω ü§≤ –Ω–∞ *${msgObj.effect}* \\[ ${expString}]`
    }
    this.write(data);
  }

  /**
   * –§—É–Ω–∫—Ü–∏—è –æ—Ç–ø—Ä–∞–≤–∫–∏ —Å–æ–æ–±—â–µ–Ω–∏–π –≤ Game
   * @param {string} data –æ–±—å–µ–∫—Ç —Å–æ–æ–±—â–µ–Ω–∏—è
   */
  write(data) {
    this.emit('BattleLog', data);
  }
}

module.exports = BattleLog;
