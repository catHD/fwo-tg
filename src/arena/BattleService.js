const { Markup } = require('telegraf');
const channelHelper = require('../helpers/channelHelper');
const arena = require('./index');

/**
 * Заказываем в три этапа:
 * 1. Выбор умения (action_{attack})
 * 2. Выбор цели ({attack}_{target})
 * 3. Выбор силы ({attack}_{target}_{proc})
 * @typedef {import ('./PlayerService').default} Player
 * @typedef {import ('./GameService')} Game
 * @typedef {import('telegraf/typings/markup').CallbackButton} CallbackButton
 */

const getActions = () => ({ ...arena.actions, ...arena.skills, ...arena.magics });

const getText = {
  order: (proc) => `У тебя осталось *${proc}%*
Заказы:\n`,
  proc: (displayName, target) => `Выбери силу _${displayName}_ на ${target ? `игрока ${target}` : 'себя'}`,
  target: (displayName) => `Выбери цель для _${displayName}_`,
};
/**
 * @param {Game} game
 * @param {import ('./OrderService').order} order
 */
const createOrder = (game, order) => {
  game.orders.orderAction(order);
};

/**
 * @param {string} charId
 * @param {Game} game
 */
const getCurrentOrders = (charId, game) => game.orders.getPlayerOrders(charId)
  .map((o) => {
    if (o.target === charId) {
      return `_${getActions()[o.action].displayName}_ (*${o.proc}%*) на *себя*`;
    }
    return `_${getActions()[o.action].displayName}_ (*${o.proc}%*) на игрока *${game.players[o.target].nick}*`;
  })
  .join('\n');

/**
 * @param {string} charId
 * @param {Game} game
 * @param {string} action
 * @returns {CallbackButton[]}
 */
const getTargetKeyboard = (charId, game, action) => {
  const player = game.players[charId];
  const { orderType } = getActions()[action];
  const proc = arena.skills[action] ? `_${arena.skills[action].proc}` : '';
  return game.alivePlayers
    .filter((target) => (orderType === 'enemy' ? !game.isPlayersAlly(player, target) : true))
    .map(({ nick, id }) => Markup.callbackButton(nick, `${action}_${id}${proc}`));
};

/**
 * Возвращает кнопки с процентом заказа
 * @param {string} action
 * @param {string} target
 * @param {number} proc
 * @returns {CallbackButton[]}
 */
const getProcentKeyboard = (action, target, proc) => Array
  .from(new Set([5, 10, 25, 50, 75, proc]))
  .filter((key) => key <= proc)
  .map((key) => Markup.callbackButton(key.toString(), `${action}_${target}_${key}`));

/**
 * Сообщение для первого этапа
 * @param {string} charId
 * @param {Game} game
 * @returns {[string, CallbackButton[][]]}
 */
const orderMessage = (charId, game) => {
  const player = game.players[charId];
  const message = getText.order(player.proc)
    .concat(getCurrentOrders(charId, game));
  const keyboard = channelHelper.getOrderButtons(player);
  return [message, keyboard];
};

/**
 * Сообщение для второго этапа
 * @param {string} charId
 * @param {Game} game
 * @param {string} action
 * @returns {[string, CallbackButton[]]}
 */
const targetMessage = (charId, game, action) => {
  const { displayName } = getActions()[action];
  const message = getText.target(displayName);
  const keyboard = getTargetKeyboard(charId, game, action);
  return [message, keyboard];
};

/**
 * Сообщение для третьего этапа
 * @param {string} charId
 * @param {Game} game
 * @param {string} action
 * @param {string} targetId
 * @returns {[string, CallbackButton[]]}
 */
const percentMessage = (charId, game, action, targetId) => {
  const { proc } = game.players[charId];
  const { nick } = game.players[targetId];
  const { displayName } = getActions()[action];
  const message = getText.proc(displayName, targetId === charId ? false : nick);
  const keyboard = getProcentKeyboard(action, targetId, proc);
  return [message, keyboard];
};

module.exports = {
  /**
   * Обработка выбранного действия (первый этап)
   * @param {string} charId
   * @param {Game} game
   * @param {string} action
   * @return {[string, CallbackButton[] | CallbackButton[][]]}
   */
  handleAction(charId, game, action) {
    if (action === 'repeat') return this.repeatOrder(charId, game);
    if (action === 'reset') return this.resetOrder(charId, game);

    const { orderType } = getActions()[action];
    const proc = arena.skills[action] ? arena.skills[action].proc : undefined;

    // Если умение направлено на себя и имеет процент (умения), то делаем заказ сразу
    if (orderType === 'self') {
      if (typeof proc !== 'undefined') {
        createOrder(game, {
          initiator: charId, target: charId, action, proc,
        });
        return orderMessage(charId, game);
      }
      return percentMessage(charId, game, action, charId);
    }
    return targetMessage(charId, game, action);
  },
  /**
   * Обработка выбранной цели (второй этап)
   * @param {string} charId
   * @param {Game} game
   * @param {string} action
   * @param {string} target
   */
  handleTarget(charId, game, action, target) {
    const proc = arena.skills[action] ? arena.skills[action].proc : undefined;

    if (typeof proc !== 'undefined') {
      createOrder(game, {
        initiator: charId, target, action, proc,
      });
      return orderMessage(charId, game);
    }
    return percentMessage(charId, game, action, target);
  },
  /**
   * Обработка выбранных процентов (третий этап)
   * @param {string} charId
   * @param {Game} game
   * @param {string} action
   * @param {string} target
   * @param {number} proc
   */
  handlePercent(charId, game, action, target, proc) {
    createOrder(game, {
      initiator: charId, target, action, proc,
    });
    return orderMessage(charId, game);
  },
  /**
   * @param {string} charId
   * @param {Game} game
   */
  repeatOrder(charId, game) {
    game.orders.repeatLastOrder(charId);
    return orderMessage(charId, game);
  },
  /**
   * @param {string} charId
   * @param {Game} game
   */
  resetOrder(charId, game) {
    game.orders.resetOrdersForPlayer(charId);
    return orderMessage(charId, game);
  },
};
