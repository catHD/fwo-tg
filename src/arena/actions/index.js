const attack = require('./attack');
const protect = require('./protect');
const regen = require('./regen');
const handsHeal = require('./handsHeal');

module.exports = {
  attack,
  protect,
  regen,
  handsHeal,
};
