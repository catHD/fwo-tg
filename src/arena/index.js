module.exports = {
  /** @type {import ('./MatchMakingService')} */
  mm: null,
  /** @type Object<string, import ('./CharacterService')> */
  characters: {},
  /** @type Object<string, import ('./GameService').default> */
  games: {},
  /**
   * @type Record<import ('../models/item').Item['code'], import ('../models/item').Item>
   * @description Item from DB
   */
  items: {},

  /** @type import ('./magics')  */
  magics: null,
  /** @type import ('./skills/') */
  skills: null,
  /** @type import ('./actions') */
  actions: null,
  /** @type {Object<string, import ('../models/clan').ClanDocument>} */
  clans: {},
};
