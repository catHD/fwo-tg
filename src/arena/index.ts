import type { Telegraf } from 'telegraf';
import type { BotContext } from '@/fwo';
import type { Clan } from '@/models/clan';
import type * as actions from './actions';
import type { CharacterService } from './CharacterService';
import type GameService from './GameService';
import type * as magics from './magics';
import type MatchMaking from './MatchMakingService';
import type * as skills from './skills';
import type { Item } from '@/schemas/item';

export default {
  mm: null as unknown as typeof MatchMaking,
  characters: {} as Record<string, CharacterService>,
  games: {} as Record<string, GameService>,
  /**
   * @description Item from DB
   */
  items: {} as unknown as Record<string, Item>,
  magics: null as unknown as typeof magics,
  skills: null as unknown as typeof skills,
  actions: null as unknown as typeof actions,
  clans: new Map<string, Clan>(),
  bot: null as unknown as Telegraf<BotContext>,
};
