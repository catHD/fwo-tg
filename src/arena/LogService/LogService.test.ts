import { LogService } from '@/arena/LogService';
import TestUtils from '@/utils/testUtils';

// npm t src/arena/BattleLog/BattleLog.test.ts

describe('BattleLog', () => {
  let messages: string[] = [];
  const writer = (data: string[]) => {
    messages = data;
  };

  const logService = new LogService(writer);

  afterEach(() => {
    messages = [];
  });

  it('should format magic messages', async () => {
    await logService.sendBattleLog([{
      actionType: 'magic',
      action: 'Magic',
      initiator: 'Player 1',
      target: 'Player 2',
      exp: 25,
    }, {
      actionType: 'magic',
      action: 'Magic',
      initiator: 'Player 1',
      target: 'Player 2',
      exp: 25,
      effect: 10,
    },
    {
      actionType: 'magic',
      action: 'Magic',
      initiator: 'Player 1',
      target: 'Player 2',
      message: 'GOD_FAIL',
    }]);

    expect(messages).toMatchSnapshot();
  });

  it('should format skill messages', async () => {
    await logService.sendBattleLog([{
      actionType: 'skill',
      action: 'Skill',
      initiator: 'Player 1',
      target: 'Player 2',
      exp: 25,
    }, {
      actionType: 'skill',
      action: 'Skill',
      initiator: 'Player 1',
      target: 'Player 2',
      message: 'SKILL_FAIL',
    }]);

    expect(messages).toMatchSnapshot();
  });

  it('should format phys messages', async () => {
    const weapon = await TestUtils.getWeapon();

    await logService.sendBattleLog([{
      actionType: 'phys',
      action: 'Phys',
      initiator: 'Player 1',
      target: 'Player 2',
      weapon,
      dmgType: 'physical',
      exp: 25,
      dmg: 10,
      hp: 90,
    }, {
      actionType: 'phys',
      action: 'Phys',
      initiator: 'Player 1',
      target: 'Player 2',
      weapon,
      message: 'DEF',
    }]);

    expect(messages).toMatchSnapshot();
  });

  it('should format dmg-magic messages', async () => {
    await logService.sendBattleLog([{
      actionType: 'dmg-magic',
      action: 'Dmg Magic',
      initiator: 'Player 1',
      target: 'Player 2',
      dmgType: 'acid',
      exp: 25,
      dmg: 10,
      hp: 90,
    },
    {
      actionType: 'dmg-magic',
      action: 'Phys',
      initiator: 'Player 1',
      target: 'Player 2',
      message: 'NO_MANA',
    }]);

    expect(messages).toMatchSnapshot();
  });

  it('should format heal-magic messages', async () => {
    await logService.sendBattleLog([{
      actionType: 'heal-magic',
      action: 'Heal Magic',
      initiator: 'Player 1',
      target: 'Player 2',
      exp: 10,
      hp: 110,
      effect: 10,
    }, {
      actionType: 'heal-magic',
      action: 'Heal Magic',
      initiator: 'Player 1',
      target: 'Player 2',
      message: 'SILENCED',
    }]);

    expect(messages).toMatchSnapshot();
  });

  it('should format heal messages with rigth order', async () => {
    await logService.sendBattleLog([{
      actionType: 'heal',
      action: 'Heal',
      initiator: 'Player 1',
      target: 'Player 2',
      expArr: [{
        id: '1', name: 'Player 1', exp: 10, val: 10,
      }],
      hp: 110,
      effect: 10,
    }, {
      actionType: 'heal',
      action: 'Heal',
      initiator: 'Player 3',
      target: 'Player 2',
      expArr: [{
        id: '3', name: 'Player 3', exp: 10, val: 10,
      }],
      hp: 120,
      effect: 10,
    }, {
      actionType: 'heal',
      action: 'Heal',
      initiator: 'Player 1',
      target: 'Player 2',
      message: 'SILENCED',
    }]);

    expect(messages).toMatchSnapshot();
  });

  it('should format long magic messages with rigth order', async () => {
    await logService.sendBattleLog([{
      actionType: 'magic-long',
      action: 'Magic 1',
      initiator: 'Player 2',
      target: 'Player 1',
      exp: 25,
    }, {
      actionType: 'magic-long',
      action: 'Magic 1',
      initiator: 'Player 1',
      target: 'Player 2',
      exp: 25,
    }, {
      actionType: 'magic-long',
      action: 'Magic 1',
      initiator: 'Player 1',
      target: 'Player 2',
      exp: 25,
    }, {
      actionType: 'magic-long',
      action: 'Magic 2',
      initiator: 'Player 1',
      target: 'Player 2',
      exp: 25,
    }, {
      actionType: 'magic-long',
      action: 'Magic 2',
      initiator: 'Player 2',
      target: 'Player 1',
      exp: 25,
    }]);

    expect(messages).toMatchSnapshot();
  });

  it('should format long dmg magic messages with rigth order', async () => {
    await logService.sendBattleLog([{
      actionType: 'dmg-magic-long',
      action: 'Magic 1',
      initiator: 'Player 2',
      target: 'Player 1',
      dmgType: 'clear',
      exp: 25,
      dmg: 25,
      hp: 75,
    }, {
      actionType: 'dmg-magic-long',
      action: 'Magic 1',
      initiator: 'Player 1',
      target: 'Player 2',
      dmgType: 'clear',
      exp: 25,
      dmg: 25,
      hp: 75,
    }, {
      actionType: 'dmg-magic-long',
      action: 'Magic 1',
      initiator: 'Player 1',
      target: 'Player 2',
      dmgType: 'clear',
      exp: 25,
      dmg: 25,
      hp: 50,
    }, {
      actionType: 'dmg-magic-long',
      action: 'Magic 2',
      initiator: 'Player 1',
      target: 'Player 2',
      dmgType: 'clear',
      exp: 25,
      dmg: 25,
      hp: 25,
    }, {
      actionType: 'dmg-magic-long',
      action: 'Magic 2',
      initiator: 'Player 2',
      target: 'Player 1',
      dmgType: 'clear',
      exp: 25,
      dmg: 25,
      hp: 50,
    }]);

    expect(messages).toMatchSnapshot();
  });

  it('should use custom message', async () => {
    await logService.sendBattleLog([{
      actionType: 'skill',
      action: 'Skill',
      exp: 10,
      initiator: 'Initiator',
      target: 'Target',
      msg: (args) => `custom ${args.initiator}|${args.target}|${args.action}|${args.actionType}`,
    }]);

    expect(messages).toMatchSnapshot();
  });
});
