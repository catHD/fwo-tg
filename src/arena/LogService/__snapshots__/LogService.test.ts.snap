// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BattleLog should format dmg-magic messages 1`] = `
[
  "*Player 1* —Å–æ—Ç–≤–æ—Ä–∏–ª _Dmg Magic_ –Ω–∞ *Player 2* –Ω–∞–Ω–µ—Å—è 10 —É—Ä–æ–Ω–∞
\\[ ‚ò£ üíî-10/90 üìñ25 ]",
  "*Player 1* –ø—ã—Ç–∞–ª—Å—è –∏—Å–ø–æ–ª—å–∑–æ–≤–∞—Ç—å _Phys_ –Ω–∞ *Player 2, –Ω–æ –Ω–µ —Ö–≤–∞—Ç–∏–ª–æ –º–∞–Ω—ã",
]
`;

exports[`BattleLog should format heal messages with rigth order 1`] = `
[
  "–ò–≥—Ä–æ–∫ *Player 2* –±—ã–ª –≤—ã–ª–µ—á–µ–Ω ü§≤ –Ω–∞ *üíñ20*
\\[ Player 1: üíñ10/110 üìñ10, Player 3: üíñ10/110 üìñ10 ]",
  "*Player 1* –ø—ã—Ç–∞–ª—Å—è —Å–æ—Ç–≤–æ—Ä–∏—Ç—å _Heal_, –Ω–æ –ø–æ–ø—ã—Ç–∫–∞ –ø—Ä–æ–≤–∞–ª–∏–ª–∞—Å—å (–±–µ–∑–º–æ–ª–≤–∏–µ)",
]
`;

exports[`BattleLog should format heal-magic messages 1`] = `
[
  "*Player 1* –∏—Å–ø–æ–ª—å–∑–æ–≤–∞–ª _Heal Magic_ –Ω–∞ *Player 2*
\\[ ‚ù§Ô∏è+10/110 üìñ10 ]",
  "*Player 1* –ø—ã—Ç–∞–ª—Å—è —Å–æ—Ç–≤–æ—Ä–∏—Ç—å _Heal Magic_, –Ω–æ –ø–æ–ø—ã—Ç–∫–∞ –ø—Ä–æ–≤–∞–ª–∏–ª–∞—Å—å (–±–µ–∑–º–æ–ª–≤–∏–µ)",
]
`;

exports[`BattleLog should format long dmg magic messages with rigth order 1`] = `
[
  "*Player 2* —Å–æ—Ç–≤–æ—Ä–∏–ª _Magic 1_ –Ω–∞ *Player 1* –Ω–∞–Ω–µ—Å—è 25 —É—Ä–æ–Ω–∞
\\[  üíî-25/75 üìñ25 ]",
  "*Player 1* —Å–æ—Ç–≤–æ—Ä–∏–ª _Magic 1_ –Ω–∞ *Player 2* –Ω–∞–Ω–µ—Å—è 50 —É—Ä–æ–Ω–∞
\\[  üíî-50/50 üìñ50 ]",
  "*Player 1* —Å–æ—Ç–≤–æ—Ä–∏–ª _Magic 2_ –Ω–∞ *Player 2* –Ω–∞–Ω–µ—Å—è 25 —É—Ä–æ–Ω–∞
\\[  üíî-25/25 üìñ25 ]",
  "*Player 2* —Å–æ—Ç–≤–æ—Ä–∏–ª _Magic 2_ –Ω–∞ *Player 1* –Ω–∞–Ω–µ—Å—è 25 —É—Ä–æ–Ω–∞
\\[  üíî-25/50 üìñ25 ]",
]
`;

exports[`BattleLog should format long magic messages with rigth order 1`] = `
[
  "*Player 2* –∏—Å–ø–æ–ª—å–∑–æ–≤–∞–ª _Magic 1_ –Ω–∞ *Player 1*
\\[ üìñ25 ]",
  "*Player 1* –∏—Å–ø–æ–ª—å–∑–æ–≤–∞–ª _Magic 1_ –Ω–∞ *Player 2*
\\[ üìñ50 ]",
  "*Player 1* –∏—Å–ø–æ–ª—å–∑–æ–≤–∞–ª _Magic 2_ –Ω–∞ *Player 2*
\\[ üìñ25 ]",
  "*Player 2* –∏—Å–ø–æ–ª—å–∑–æ–≤–∞–ª _Magic 2_ –Ω–∞ *Player 1*
\\[ üìñ25 ]",
]
`;

exports[`BattleLog should format magic messages 1`] = `
[
  "*Player 1* –∏—Å–ø–æ–ª—å–∑–æ–≤–∞–≤ _Magic_ –Ω–∞ *Player 2*
\\[ üìñ25 ]",
  "*Player 1* –∏—Å–ø–æ–ª—å–∑–æ–≤–∞–≤ _Magic_ –Ω–∞ *Player 2* —Å —ç—Ñ—Ñ–µ–∫—Ç–æ–º 10
\\[ üìñ25 ]",
  "–ó–∞–∫–ª–∏–Ω–∞–Ω–∏–µ _Magic_ *Player 1* –ø—Ä–æ–≤–∞–ª–∏–ª–æ—Å—å –ø–æ –≤–æ–ª–µ –±–æ–≥–æ–≤",
]
`;

exports[`BattleLog should format phys messages 1`] = `
[
  "*Player 1* —Ä—É–±–∞–Ω—É–ª *Player 2* _–≠–ø–∏—á–µ—Å–∫–∏–º –ú–µ—á–æ–º_ –∏ –Ω–∞–Ω—ë—Å *10* —É—Ä–æ–Ω–∞
\\[ üíî-10/90 üìñ25 ]",
  "*Player 1* –∞—Ç–∞–∫–æ–≤–∞–ª *Player 2* _–≠–ø–∏—á–µ—Å–∫–∏–º –ú–µ—á–æ–º_, –Ω–æ —Ç–æ—Ç —Å–º–æ–≥ –∑–∞—â–∏—Ç–∏—Ç—å—Å—è \\[]",
]
`;

exports[`BattleLog should format skill messages 1`] = `
[
  "*Player 1* –∏—Å–ø–æ–ª—å–∑–æ–≤–∞–ª _Skill_ –Ω–∞ *Player 2*
\\[ üìñ25 ]",
  "*Player 1* –ø—ã—Ç–∞–ª—Å—è –∏—Å–ø–æ–ª—å–∑–æ–≤–∞—Ç—å —É–º–µ–Ω–∏–µ _Skill_, –Ω–æ —É –Ω–µ–≥–æ –Ω–µ –≤—ã—à–ª–æ",
]
`;

exports[`BattleLog should use custom message 1`] = `
[
  "custom Initiator|Target|Skill|skill
\\[ üìñ10 ]",
]
`;
